### PR TITLE
Support for IE 9+

### PR DESCRIPTION
**Pull request by mkkeck**
Adds compatibility for IE9, IE10 and IE11 #121

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -41,14 +43,21 @@
     "preact": "^10.0.4",
     "purify-css": "^1.2.5",
     "purifycss-webpack": "^0.7.0",
+    "react": "^16.12.0",
+    "react-dom": "^16.12.0",
```
Comment (jaller94) [inline package.json]: What's the reason for adding React?
The project uses Preact/compat.

File: package.json [hunk 2]
```diff
@@ -41,14 +43,21 @@
...
+    "utils": "^0.3.1",
     "webpack": "^4.41.2",
     "webpack-bundle-size-analyzer": "^3.1.0",
     "webpack-cli": "^3.3.10",
     "webpack-dev-server": "^3.9.0",
     "webpack-hot-middleware": "^2.25.0",
     "yaml-loader": "^0.5.0"
+  },
+  "dependencies": {
+    "currentscript": "^1.1.1"
```
Comment (jaller94) [inline package.json]: Was this missing? I don't see `currentscript` being used in the new code.

File: babel.config.js [hunk 3]
```diff
@@ -1,7 +1,13 @@
 // This is the configuration file for babel
 /* eslint-env node */
 module.exports = {
-    presets: ["@babel/preset-env", "@babel/preset-react"],
+    presets: [
+        ["@babel/preset-env", {
+            useBuiltIns: "entry",
```
Comment (jaller94) [inline babel.config.js]: In this other ticket @leonexcc was able to use "usage", which allowed to not add imports to the code and worry less about keeping polyfills up-to-date. The compatibility there is based on browserlist which I think is preferred while we don't have a lot of time for Klaro's development.

https://github.com/KIProtect/klaro/pull/124/commits/885347481eac466dbc1d68a6d64028435d525c02

File: src/consent-manager.js [hunk 4]
```diff
@@ -123,9 +128,10 @@ export default class ConsentManager {
     }
 
     saveConsents(){
-        if (this.consents === null)
+        var v = this.consents
+        if (v === null)
             deleteCookie(this.cookieName)
-        const v = JSON.stringify(this.consents)
+        v = v !== null ? JSON.stringify(this.consents) : ""
```
Comment (jaller94) [inline src/consent-manager.js]: This reassignment of `v` seems a bit too wild.
A variable should only have one type (and a speaking name, while you're at it).

File: src/consent-manager.js [hunk 5]
```diff
@@ -144,7 +150,7 @@ export default class ConsentManager {
                 continue
             this.updateAppElements(app, consent)
             this.updateAppCookies(app, consent)
-            if (app.callback !== undefined)
+            if (typeof app.callback === "function")
```
Comment (jaller94) [inline src/consent-manager.js]: Great job for being more verbose where it makes sense. 👍

File: src/consent-manager.js [hunk 6]
```diff
@@ -239,7 +263,7 @@ export default class ConsentManager {
             return
 
         function escapeRegexStr(str) {
-            return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+            return str.replace(/[\-\[\]\/{}()*+?.\\^$|]/g, "\\$&");
```
Comment (jaller94) [inline src/consent-manager.js]: What's changing here?
Would you mind explaining the new regex?
Why did some characters get removed and others added?

File: src/utils/i18n.js [hunk 7]
```diff
@@ -62,10 +62,22 @@ function hget(d, key, defaultValue){
...
+
+    // try to get first language as fallback
+    if (value === undefined) {
+        for (fb of trans.keys()) {
+            value = hget(trans, [fb, ...kl]);
+            if (value !== undefined) {
+                break;
+            }
+        }
+    }
```
Comment (jaller94) [inline src/utils/i18n.js]: Can we instead call `t` with the first language as a parameter?
Also, Is this change related to the IE9 compatibility?

File: src/consent-manager.js (hunk 6 above)
Comment (mkkeck) [reply src/consent-manager.js]: Regexp in [] are not needed to be escaped, my IDE (phpStorm) as optimized `redudant character escapings`

File: src/utils/i18n.js (hunk 7 above)
Comment (mkkeck) [reply src/utils/i18n.js]: I think not, do you really know, that 'en' for example is the first language? What if an sentance is missing in an language?

File: src/consent-manager.js (hunk 4 above)
Comment (mkkeck) [reply src/consent-manager.js]: This fix is for older browsers: JSON.parse / JSON.stringify en empty strings results in a syntax error (tested on IE 9). But I see: `const` should be replaced with `var?

File: package.json (hunk 2 above)
Comment (mkkeck) [reply package.json]: Don't really know, but I think it is required `current-executing-script`

File: package.json (hunk 1 above)
Comment (mkkeck) [reply package.json]: I see would be removed

File: src/utils/i18n.js (hunk 7 above)
Comment (jaller94) [reply src/utils/i18n.js]: Right, we shouldn't assume 'en' to be the default for the website owner's content.

File: src/utils/i18n.js [hunk 8]
```diff
@@ -62,10 +62,22 @@ function hget(d, key, defaultValue){
...
+    var fb
     if (!Array.isArray(kl))
         kl = [kl]
-    const value = hget(trans, [lang, ...kl])
-    if (value === undefined){
+    var value = hget(trans, [lang, ...kl])
+
+    // try to get first language as fallback
+    if (value === undefined) {
+        for (fb of trans.keys()) {
```
Comment (jaller94) [inline src/utils/i18n.js]: `fb` stands for `fallback`?

File: src/utils/i18n.js [hunk 9]
```diff
@@ -62,10 +62,22 @@ function hget(d, key, defaultValue){
...
 export function t(trans, lang, key, ...params){
     var kl = key
+    var fb
     if (!Array.isArray(kl))
         kl = [kl]
-    const value = hget(trans, [lang, ...kl])
-    if (value === undefined){
+    var value = hget(trans, [lang, ...kl])
+
+    // try to get first language as fallback
```
Comment (jaller94) [inline src/utils/i18n.js]: This comment might be outdated already as the code seems to look through all translations to find a fallback.

File: babel.config.js (hunk 3 above)
Comment (mkkeck) [reply babel.config.js]: How does this work? Should the user add polyfills by himself or are all required polifills added automatic?

File: src/utils/i18n.js (hunk 8 above)
Comment (mkkeck) [reply src/utils/i18n.js]: yes

File: src/utils/i18n.js (hunk 9 above)
Comment (mkkeck) [reply src/utils/i18n.js]: But is does exit on the first match :)

File: src/utils/i18n.js (hunk 7 above)
Comment (mkkeck) [reply src/utils/i18n.js]: Perhabs it would be a good idea to use something like `defaults` as default language which must be filled with all needed strings. Then each translation is moved to config, so the base could be smaller. For me, but that's my personal opinion, it does not make sence to download an app with dozens of languages included. The other point why should I translate things, which must not.

File: package.json (hunk 2 above)
Comment (mkkeck) [reply package.json]: fixed with: https://github.com/mkkeck/coi/commit/f834bb0f0e6b18f91552cd99063337d56f1b97ec

File: package.json (hunk 1 above)
Comment (mkkeck) [reply package.json]: fixed with: mkkeck@f834bb0

File: src/consent-manager.js (hunk 4 above)
Comment (mkkeck) [reply src/consent-manager.js]: fixed with: https://github.com/mkkeck/coi/commit/bf9bcd4cf5546bdebd8344c377c6ccb18fc7644b

File: src/utils/i18n.js (hunk 7 above)
Comment (mkkeck) [reply src/utils/i18n.js]: https://github.com/mkkeck/coi/commit/bf9bcd4cf5546bdebd8344c377c6ccb18fc7644b

File: src/utils/i18n.js (hunk 9 above)
Comment (mkkeck) [reply src/utils/i18n.js]: https://github.com/mkkeck/coi/commit/bf9bcd4cf5546bdebd8344c377c6ccb18fc7644b

**Comment by adewes** [comment]
I just made the necessary changes to make Klaro IE10 & IE11 (and possibly IE9 though I didn't test that yet) compatible.
